### ISSUE TITLE
Validate best-checkpoint state across trainer resume

### ISSUE THREAD

**Issue by shapovalovei** [shapovalovei/YOLOX-OBB#48]
## Phase

Bounded validation only. Implementation is NOT authorized by this Issue.

## Repository baseline

Repository: `shapovalovei/YOLOX-OBB`

Current main: `6ddaed5d5d6103b1de11d14cf3e65167b84c8c73`

Behavioral baseline: `0695e2834ff5716dfd51fcd6ac7b2300b7bc27ce`

Validation must preserve the maintained OBB contract and must not change tracked files.

## Global constraints

- no source, test, configuration, or documentation changes;
- no branch, commit, PR, or CI;
- no training, optimizer steps, dataset changes, or model-quality claims;
- temporary probes/checkpoints may exist only outside the repository;
- separate FACT, INFERENCE, and UNVERIFIED;
- owner/orchestrator reviews and closes this Issue.

## Local hypothesis

The maintained trainer saves `start_epoch`, model state, optimizer state, and optional AMP state, but does not persist the historical best metric. Resume may therefore reset `best_ap` and allow a weaker post-resume evaluation to replace the true best checkpoint.

A temporary checkpoint probe stored `best_ap=0.8`; local resume restored the epoch but produced `best_ap=0.0`. This must be validated through the maintained trainer methods.

## Validation matrix

Use a temporary directory and a mocked/synthetic trainer lifecycle. No optimizer step or model training is permitted.

1. Create a checkpoint containing model, optimizer, optional AMP state, `start_epoch`, and a prior best metric.
2. Resume through the actual `resume_train` path.
3. Verify epoch, optimizer, AMP, and best-metric state.
4. Run mocked evaluation below the stored best.
5. Run mocked evaluation above the stored best.
6. Inspect latest and best checkpoint save decisions and serialized metadata.
7. Repeat without prior best metadata for legacy compatibility.

## Acceptance / falsification

**A — acceptable:** prior best state survives resume; lower metrics do not replace the best checkpoint; higher metrics update it; missing legacy metadata has defined safe behavior.

**B — defect proven:** best state resets, a weaker metric overwrites the best checkpoint, or resume changes save selection unexpectedly.

**C — inconclusive:** mocked lifecycle cannot reach the maintained save/resume path or required checkpoint fields are unavailable.

## Stop conditions

No real training, optimizer update, dataset access, tracked checkpoint artifact, or implementation is authorized.

## Completion handoff

Post temporary checkpoint schema, mocked lifecycle trace, save decisions, exact results, FACT / INFERENCE / UNVERIFIED, verdict A/B/C, and future implementation acceptance criteria.

**Comment by shapovalovei** [comment]
Orchestrator baseline update before validation: PR #55 / Issue #54 has been completed and squash-merged. Current `origin/main` is now `7bd6ddb92c977fc876a83fc567d4dea98fa68156`. The #54 change is limited to optional `save_interval` handling in `Trainer.after_epoch()` and does not change `best_ap` persistence or resume semantics. Execute #48 validation against current `origin/main`; keep the existing #48 hypothesis and scope unchanged.

**Comment by shapovalovei** [comment]
## Validation handoff

### Scope and baseline

- Repository: `shapovalovei/YOLOX-OBB`.
- Issue: #48, validated against detached `origin/main` at exact SHA `7bd6ddb92c977fc876a83fc567d4dea98fa68156` (`fix: handle absent save_interval in trainer (#55)`).
- The orchestrator baseline comment on this Issue was read and applied. Issue #54 / PR #55 was not used as evidence for or against `best_ap` persistence; the probe used a nonmatching `save_interval` only to keep periodic-save behavior out of the metric decision.
- Environment: macOS `Darwin 25.6.0`, x86_64; `/Users/eugene/creditCardReader/venv/bin/python`; Python `3.11.11`; PyTorch `2.2.2`; `torch.cuda.is_available() == False`; pytest was unavailable.
- The probe ran from a clean detached temporary worktree at the baseline. No real training, optimizer step, dataset access, GPU work, tracked-file modification, branch, commit, PR, Issue, or CI was used. Checkpoints existed only in a temporary `/tmp` directory.

### Synthetic checkpoint schemas

Historical-best input (`latest_ckpt.pth`):

```text
start_epoch, model, optimizer, amp, best_ap=0.8, best_ap_available=true
```

Legacy input (`latest_ckpt.pth`):

```text
start_epoch, model, optimizer, amp
```

The maintained `Trainer.save_ckpt()` output observed for `latest`, `last_epoch`, and `best_ckpt` was:

```text
start_epoch, model, optimizer, amp
```

Neither `best_ap` nor `best_ap_available` was serialized by the maintained save path.

### Resume trace

The actual maintained `Trainer.resume_train(model)` path was called with `resume=True`, an explicit temporary `latest_ckpt.pth`, and no `start_epoch` override. For the historical-best input it restored:

```text
start_epoch = 8
model weight digest = ad02908e7dc8436b  # source weight was all 2.0
optimizer lr = 0.123
optimizer momentum = 0.9
optimizer state entries = 2
AMP load = {amp_marker: prior_amp, loss_scale: 128.0}
```

The resume logger trace was `resume training` followed by `loaded checkpoint 'True' (epoch 8)`. After resume, trainer metric state was:

```text
best_ap = 0
best_ap_available = false
```

Thus model, optimizer, epoch, and optional AMP state survived the maintained resume path, while the historical best state did not.

### Mocked evaluation and save decisions

Each case then called the actual `Trainer.after_epoch()` lifecycle with mocked `exp.eval()` and actual maintained save logic. Epoch 8 produced `start_epoch=9` in saved checkpoints.

Historical best `0.8`, mocked evaluation `0.7`:

```text
latest:     is_best=false, start_epoch=9
last_epoch: is_best=true,  start_epoch=9
best_ckpt:  overwritten; serialized keys were start_epoch, model, optimizer, amp
trainer after evaluation: best_ap=0.7, best_ap_available=true
```

The pre-existing synthetic `best_ckpt.pth` SHA-256 was `17d7dbbda70661cbe306eb87fec63a4dd41b39fe7bad2be8eb7089975d7156a3`; it changed after the lower metric. This is the required weaker-evaluation overwrite caused by the reset state.

Historical best `0.8`, mocked evaluation `0.9`:

```text
latest:     is_best=false, start_epoch=9
last_epoch: is_best=true,  start_epoch=9
best_ckpt:  overwritten as expected for a higher metric
trainer after evaluation: best_ap=0.9, best_ap_available=true
```

Legacy checkpoint without best-metric metadata:

```text
resume: start_epoch=8, best_ap=0, best_ap_available=false, AMP restored
latest: is_best=false, start_epoch=9
last_epoch: is_best=true for first valid mocked metric 0.7
best_ckpt: created; no best-metric metadata serialized
trainer after evaluation: best_ap=0.7, best_ap_available=true
```

### FACT / INFERENCE / UNVERIFIED

**FACT**

- `Trainer.__init__` initializes `best_ap=0` and `best_ap_available=False`.
- `Trainer.resume_train()` loads `model`, `optimizer`, optional `amp`, and `start_epoch`, but does not read `best_ap` or `best_ap_available` from the checkpoint.
- `Trainer.save_ckpt()` builds checkpoint state without `best_ap` or `best_ap_available`.
- `evaluate_and_save_model()` decides `is_best` before updating the in-memory metric state, so the reset `best_ap=0` makes `0.7 > 0` and copies the weaker `last_epoch` checkpoint to `best_ckpt.pth`.
- The synthetic lifecycle reached the maintained resume, evaluation, and save methods without an optimizer step or training.

**INFERENCE**

- The smallest causal persistence/resume boundary is the trainer checkpoint state contract between `Trainer.save_ckpt()` and `Trainer.resume_train()`: the best-metric fields are neither persisted by the writer nor restored by the reader. The comparison and copy mechanism itself behaves consistently with the in-memory value it receives.
- A backward-compatible implementation should persist and restore `best_ap` and `best_ap_available` for new checkpoints. For legacy checkpoints missing either field, resume should remain non-failing and retain the current safe defaults (`best_ap=0`, `best_ap_available=false`); because historical best is unknowable, the first valid post-resume metric may become best. No historical metric should be fabricated.

**UNVERIFIED**

- Real GPU/Apex AMP behavior, distributed execution, EMA-specific save behavior, and model-quality impact were not validated and are outside this bounded synthetic task.
- No conclusion is drawn about the optional `save_interval` change in #54/#55.

### Verdict

**B — DEFECT PROVEN**

### Recommended next phase

Create or authorize a separate generic YOLOX-OBB implementation Issue (not created by this validation) for the checkpoint state contract. Its acceptance criteria should include: new-checkpoint persistence and resume restoration of both best-metric fields; lower post-resume metrics do not update/copy `best_ckpt`; higher metrics do; legacy checkpoints without metadata resume without error and use the defined first-valid-metric fallback; and focused regression coverage verifies serialized metadata and latest/best save decisions. Issue #48 remains open for orchestrator review.

**Comment by shapovalovei** [comment]
Orchestrator acceptance: verdict B accepted. Validation reproduced the maintained resume/save/evaluation path on current main and demonstrated the checkpoint-contract defect: trainer initialization resets `best_ap`/`best_ap_available`; checkpoint writing does not serialize those fields; resume does not restore them; consequently a weaker post-resume metric can be classified as best and overwrite the historical `best_ckpt`. The result is consistent with the current trainer source and is not dependent on Issue #54/#55 behavior. A separate implementation Issue will define persistence/restoration plus backward-compatible legacy-checkpoint behavior. Validation phase complete.
